daemon/prompter.c

```diff
@@ -1169,11 +1317,172 @@ prompter_disconnect(prompter_t *self)
+        self->wtc_connection = NULL;
+
+    self->wtc_prompter = NULL;
+}
+
+static watcher_t *
+watcher_create(prompter_t *prompter, GDBusConnection *connection, const gchar *name)
+{
+    watcher_t *self = g_malloc0(sizeof *self);
+    watcher_ctor(self, prompter, connection, name);
+    return self;
+}
+
+static void
+watcher_delete(watcher_t *self)
+{
+    if( self ) {
+        watcher_dtor(self);
+        g_free(self);
+    }
+}
+
+static void
+watcher_delete_cb(void *self)
+{
+    watcher_delete(self);
+}
+
+static void
+watcher_watch(watcher_t *self)
+{
+    self->wtc_watcher = g_bus_watch_name_on_connection(
+            self->wtc_connection, self->wtc_name, G_BUS_NAME_OWNER_FLAGS_NONE,
+            NULL, watcher_handle_name_lost_cb, self, NULL);
+    log_debug("watching for '%s' to leave bus", self->wtc_name);
+}
+
+static void
+watcher_unwatch(watcher_t *self)
+{
+    if( self->wtc_watcher )
+        g_bus_unwatch_name(self->wtc_watcher);
+    self->wtc_watcher = 0;
+}
+
+static void
+watcher_handle_name_lost_cb(GDBusConnection *connection, const gchar *name, gpointer aptr)
+{
+    (void)connection; // unused
+    (void)name; // unused
+    watcher_t *self = aptr;
+    log_debug("'%s' left bus", self->wtc_name);
+    watcher_notify_name_lost(self);
+}
+
+static void
+watcher_name_has_owner(watcher_t *self)
+{
+    change_cancellable_steal(&self->wtc_cancellable, g_cancellable_new());
+    g_dbus_connection_call(self->wtc_connection,
+                           DBUS_SERVICE,
+                           DBUS_PATH,
+                           DBUS_INTERFACE,
+                           DBUS_METHOD_NAME_HAS_OWNER,
+                           g_variant_new("(s)", self->wtc_name),
+                           NULL,
+                           G_DBUS_CALL_FLAGS_NONE,
+                           -1,
+                           self->wtc_cancellable,
+                           watcher_name_has_owner_cb,
+                           self);
+}
+
+static void
+watcher_name_has_owner_cb(GObject *obj, GAsyncResult *res, gpointer aptr)
+{
+    GDBusConnection *con  = G_DBUS_CONNECTION(obj);
+    GError          *err  = NULL;
+    GVariant        *rsp  = NULL;
+
+    if( !(rsp = g_dbus_connection_call_finish(con, res, &err)) ) {
+        /* NB: aptr might be invalid pointer, check before use if the call
+         *     was canceled. Thus setting self to NULL here. */
+        watcher_t *self = NULL;
+        if( err ) {
+            log_err("error reply from dbus server: domain=%u code=%d message='%s'",
+                    (unsigned)err->domain, err->code, err->message);
+            /* aptr is invalid if the call was cancelled */
+            if( err->code != G_IO_ERROR_CANCELLED )
+                self = aptr;
+        }
+        else {
+            log_err("null reply from dbus server");
+            self = aptr;
+        }
+
+        if( self )
+            change_cancellable_steal(&self->wtc_cancellable, NULL);
```

Comment:
one more matter of taste issue, feel free to ignore: there are multiple nested conditional execution paths leading to similar sets of actions - which could be avoided by a bit of rearranging, something like:

```
/* NB: aptr might be invalid pointer, check
 *     before use if the call was canceled */
watcher_t *self      = aptr;
gboolean   has_owner = true;

if( !(rsp = g_dbus_connection_call_finish(con, res, &err)) ) {
    if( err ) {
        log_err("error reply from dbus server: domain=%u code=%d message='%s'",
                (unsigned)err->domain, err->code, err->message);
        /* aptr is invalid if the call was cancelled */
        if( err->code == G_IO_ERROR_CANCELLED )
            self = NULL;
    }
}
else {
    g_variant_get(rsp, "(b)", &has_owner);
    log_debug("'%s' %s owner", self->wtc_name, (has_owner ? "has" :  "doesn't have"));
}

if( self ) {
    change_cancellable_steal(&self->wtc_cancellable, NULL);
    if( !has_owner )
        watcher_notify_name_lost(self);
}
```

Reply:
Well, yes. At the same time I wanted to protect the pointer from stupid mistakes when someone changes the code next time but maybe keeping it leaner would achieve that goal better. :thinking: